src/main/java/traffic/TrafficLightSimulator.java

```diff
@@ -0,0 +1,78 @@
+package main.java.traffic;
+
+//A computerised traffic light. 
+class TrafficLightSimulator implements Runnable {
+	private Thread thrd; // holds the thread that runs the simulation
+	private TrafficLightColour tlc; // holds the current traffic light color
+	boolean stop = false; // set to true to stop the simulation
+	boolean changed = false; // true when the light has changed
+
+	TrafficLightSimulator(TrafficLightColour init) {
+		tlc = init;
+
+		thrd = new Thread(this);
+		thrd.start();
+	}
+
+	TrafficLightSimulator() {
+		tlc = TrafficLightColour.RED;
+
+		thrd = new Thread(this);
+		thrd.start();
+	}
+
+	// Start up the light.
+	public void run() {
+		while (!stop) {
+			try {
+				Thread.sleep(tlc.getDelay());
+			} catch (InterruptedException exc) {
+				System.out.println(exc);
+			}
+			changeColor();
+		}
+	}
+
+	// Change color.
+	synchronized void changeColor() {
+		switch (tlc) {
+		case RED:
+			tlc = TrafficLightColour.RED_YELLOW;
+			break;
+		case YELLOW:
+			tlc = TrafficLightColour.RED;
+			break;
+		case GREEN:
+			tlc = TrafficLightColour.YELLOW;
+			break;
+		case RED_YELLOW:
+			tlc = TrafficLightColour.GREEN;
+			break;
+		}
```

Comment:
i had a think about it but it was too late so gave up...

wait, I mean... i did it on my own as the book instructed me to...